core/src/worker/workflow/machines/workflow_machines.rs

```diff
@@ -367,9 +367,17 @@ impl WorkflowMachines {
     /// "no work" situation. Possibly, it may know about some work the machines don't, like queries.
     pub(crate) fn get_wf_activation(&mut self) -> WorkflowActivation {
         let jobs = self.drive_me.drain_jobs();
+        // Even though technically we may have satisfied all the criteria to be done with replay,
+        // query only activations are always "replaying" to keep things sane.
+        let all_query = jobs.iter().all(|j| {
+            matches!(
+                j.variant,
+                Some(workflow_activation_job::Variant::QueryWorkflow(_))
+            )
+        });
         WorkflowActivation {
             timestamp: self.current_wf_time.map(Into::into),
-            is_replaying: self.replaying,
+            is_replaying: self.replaying || all_query,
```

Comment:
I technically agree with the idea that queries should always be considered as replay (or _never_, that's debatable, but it should be consistent), yet I'm a bit worried that this specific change may eventually end up making future bugs much harder to diagnose.

Not blocking at this time, but I think we should consider refactoring the activation protocol so that there no need for assumptions. Like have core presort jobs and attach the `is_replaying` flag at the job level, or make it so that an activation may never contain both a query and some non-query jobs.

Reply:
Yeah, I agree. I think we should do the latter option, where queries always get their own activation independent of other jobs. It might even make sense to segregate the activation type entirely.